cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueDebugger.java

```diff
@@ -0,0 +1,308 @@
+    QueueStatistics stats = new QueueStatistics();
+    int currentSection = 1;
+
+    for (Long groupId : barriers.keySet()) {
+      LOG.info("Scanning barriers for group {}", groupId);
+      Collection<QueueBarrier> groupBarriers = barriers.get(groupId);
+      Iterator<QueueBarrier> barrierIterator = groupBarriers.iterator();
+
+      QueueBarrier previous;
+      QueueBarrier current = null;
+      while (barrierIterator.hasNext()) {
+        previous = current;
+        current = barrierIterator.next();
+
+        if (previous != null) {
+          // after first iteration
+          LOG.info("Scanning section {}/{}...", currentSection, barriers.size());
+          QueueStatistics sectionStats = scanQueue(tx, queueName, previous, current);
+          LOG.info("Section results: {}", sectionStats.getReport());
+
+          stats = QueueStatistics.of(stats, sectionStats);
+          currentSection++;
+        }
+      }
+      // handle last barrier
+      LOG.info("Scanning section {}/{}...", currentSection, barriers.size());
+      stats = QueueStatistics.of(stats, scanQueue(tx, queueName, current, null));
+      LOG.info("Scanning complete", currentSection, barriers.size());
+    }
+
+    LOG.info("Total results: {}", stats.getReport());
+    stateStore.commitTx();
+  }
+
+  private QueueStatistics scanQueue(Transaction tx, QueueName queueName, QueueBarrier start,
+                                    @Nullable QueueBarrier end) throws IOException {
+
+    byte[] queueRowPrefix = QueueEntryRow.getQueueRowPrefix(queueName);
+
+    ConsumerGroupConfig consumerConfig = start.getGroupConfig();
+    LOG.info("Got consumer group config: {}", consumerConfig);
+
+    HBaseQueueAdmin admin = queueClientFactory.ensureTableExists(queueName);
+    TableId tableId = admin.getDataTableId(queueName, QueueConstants.QueueType.SHARDED_QUEUE);
+    HTable hTable = queueClientFactory.createHTable(tableId);
+
+    LOG.info("Looking at HBase table: {}", Bytes.toString(hTable.getTableName()));
+
+    byte[] stateColumnName = Bytes.add(QueueEntryRow.STATE_COLUMN_PREFIX,
+                                       Bytes.toBytes(consumerConfig.getGroupId()));
+
+
+    int distributorBuckets = queueClientFactory.getDistributorBuckets(hTable.getTableDescriptor());
+    ShardedHBaseQueueStrategy queueStrategy = new ShardedHBaseQueueStrategy(distributorBuckets);
+
+    Scan scan = new Scan();
+    scan.setStartRow(start.getStartRow());
+    if (end != null) {
+      scan.setStopRow(end.getStartRow());
+    }
+    scan.addColumn(QueueEntryRow.COLUMN_FAMILY, QueueEntryRow.DATA_COLUMN);
+    scan.addColumn(QueueEntryRow.COLUMN_FAMILY, QueueEntryRow.META_COLUMN);
+    scan.addColumn(QueueEntryRow.COLUMN_FAMILY, stateColumnName);
+    scan.setCaching(100);
+    scan.setMaxVersions(1);
+
+    LOG.info("Scanning section with scan: {}", GSON.toJson(scan));
+    QueueStatistics stats = new QueueStatistics();
+
+    ConsumerConfig consConfig = new ConsumerConfig(consumerConfig, 0);
```

Comment:
Is there a way to fetch list of instanceIds? Didn't understand what instanceId would be in this case. If groupSize were 5, then would the instanceIds be 0, 1, 2, 3, 4?